services/api-server/src/simcore_service_api_server/api/routes/files.py

```diff
@@ -1,72 +1,152 @@
-    raise NotImplementedError()
+    # Adapts storage API model to API model
+    files_metadata = deque()
+    for stored_file_meta in stored_files:
+        try:
+            assert stored_file_meta.user_id == user_id  # nosec
+            assert stored_file_meta.file_id  # nosec
 
+            meta = convert_metadata(stored_file_meta)
 
-@router.post(":upload", response_model=FileMetadata)
-async def upload_file(file: UploadFile = File(...)):
-    """Uploads a single file to the system
+        except (ValidationError, ValueError, AttributeError) as err:
+            logger.warning(
+                "Skipping corrupted entry in storage '%s' (%s)"
+                "TIP: check this entry in file_meta_data table.",
+                stored_file_meta.file_uuid,
+                err,
+            )
 
-    To upload multiple with one call, see upload_files
-    """
-    metadata = await the_fake_impl.save(file)
-    return metadata
+        else:
+            files_metadata.append(meta)
+
+    return list(files_metadata)
 
 
-async def upload_single_file_impl(
+@router.put("/content", response_model=FileMetadata)
+async def upload_file(
     file: UploadFile = File(...),
-    _storage_client: StorageApi = Depends(get_api_client(StorageApi)),
+    content_length: Optional[str] = Header(None),
+    storage_client: StorageApi = Depends(get_api_client(StorageApi)),
+    user_id: int = Depends(get_current_user_id),
 ):
-    # TODO: FileResponse automatically computes etag. See how is done
-
-    # TODO: every file uploaded is sent to S3 and a link is returned
-    # TODO: every session has a folder. A session is defined by the access token
+    """Uploads a single file to the system"""
+    # TODO: For the moment we upload file here and re-upload to S3
+    # using a pre-signed link. This is far from ideal since we are using the api-server as a
+    # passby service for all uploaded data which can be a lot.
+    # Next refactor should consider a solution that directly uploads from the client to S3
+    # avoiding the data trafic via this service
     #
 
-    # TODO: this is just a ping with retries
-    ## await storage_client.get("/")
-    raise NotImplementedError()
+    # assign file_id.
+    meta: FileMetadata = await FileMetadata.create_from_uploaded(
+        file, file_size=content_length, created_at=datetime.utcnow().isoformat()
+    )
+    logger.debug("Assigned id: %s of %s bytes", meta, content_length)
+
+    # upload to S3 using pre-signed link
+    presigned_upload_link = await storage_client.get_upload_link(
+        user_id, meta.file_id, meta.filename
+    )
 
+    logger.info("Uploading %s to %s ...", meta, presigned_upload_link)
 
-# TODO: disabled until actual use case is presented
-# @router.post(":upload-multiple", response_model=List[FileMetadata])
-#
-async def _upload_files(files: List[UploadFile] = File(...)):
+    async with httpx.AsyncClient(timeout=httpx.Timeout(5.0, write=3600)) as client:
```

Comment:
I would put the write on a env var

Reply:
But I followed [the same as here](https://github.com/ITISFoundation/osparc-simcore/blob/feaa2e243a96ef8174dd85f42ad1b5a0c3d7f9fb/services/web/server/src/simcore_service_webserver/exporter/formatters/formatter_v1.py#L34)